packages/types/src/data/abstract-query.ts

```diff
@@ -0,0 +1,240 @@
+/** The root of the abstract query */
+export interface AbstractQuery {
+	/** Marked as entrypoint of the query */
+	root: true;
+
+	/** Place where the data is stored */
+	datastore: string;
+
+	/** Name of the collection within the datastore */
+	collection: string;
+
+	/** All fields to select in the query */
+	fieldNodes: AbstractQueryFieldNode[];
+
+	// modifiers?: {
+	// 	limit?: AbstractQueryNodeLimit;
+	// 	filter?: AbstractQueryFilterNode[];
+	// };
+}
+
+type AbstractQueryNodeType = 'primitive' | 'fn' | 'm2o' | 'o2m' | 'a2o' | 'o2a';
+
+interface AbstractQueryNode {
+	/** the type of the node */
+	type: AbstractQueryNodeType;
+}
+
+/** A group of all possible field types */
+export type AbstractQueryFieldNode =
+	| AbstractQueryFieldNodePrimitive
+	| AbstractQueryFieldNodeFn
+	| AbstractQueryFieldNodeRelated;
+
+/** Generic primitive value read from the datastore field */
+export interface AbstractQueryFieldNodePrimitive extends AbstractQueryNode {
+	type: 'primitive';
+
+	/** the name of the attribute */
+	field: string;
+}
+
+export type AbstractQueryFn = 'year' | 'month' | 'week' | 'day' | 'weekday' | 'hour' | 'minute' | 'second';
```

Comment:
Can/should this type derive from `FieldFunction`? Though it seems like `AbstractQueryFn` doesn't have `count`, is that intentional (maybe this is meant as a fresh start)?

  - Reference to `FieldFunction`:

    https://github.com/directus/directus/blob/098dd359afb6e513aaf5c5d0fe0de225973eb33f/packages/types/src/fields.ts#L15

  - Reference to `FUNCTIONS` constant:

    https://github.com/directus/directus/blob/098dd359afb6e513aaf5c5d0fe0de225973eb33f/packages/constants/src/fields.ts#L69

Reply:
good point. I think `count` got lost accidentally so we could use this type definition from the constants package. or was is in fact intentional @rijkvanzanten?

Reply:
It was removed intentionally, as count isn't really a function you apply to a primitive, but rather an aggregate query you apply to a nested collection set 👍🏻